native/0004-envoy-url-socks5-param.patch

```diff
@@ -0,0 +1,101 @@
+ components/cronet/native/engine.cc      | 33 ++++++++++++++++++++++++++-------
+ components/cronet/native/sample/main.cc | 11 +++++++++--
+ 2 files changed, 35 insertions(+), 9 deletions(-)
+
+diff --git a/components/cronet/native/engine.cc b/components/cronet/native/engine.cc
+index a4e93b23c5b3f..cda95ba910509 100644
+--- a/components/cronet/native/engine.cc
++++ b/components/cronet/native/engine.cc
+@@ -15,6 +15,7 @@
+ #include "base/memory/raw_ptr.h"
+ #include "base/no_destructor.h"
+ #include "build/build_config.h"
++#include "base/strings/escape.h"
+ #include "components/cronet/cronet_context.h"
+ #include "components/cronet/cronet_global_state.h"
+ #include "components/cronet/native/generated/cronet.idl_impl_struct.h"
+@@ -24,6 +25,7 @@
+ #include "components/cronet/version.h"
+ #include "components/grpc_support/include/bidirectional_stream_c.h"
+ #include "net/base/hash_value.h"
++#include "net/base/url_util.h"
+ #include "net/url_request/url_request_context.h"
+ #include "net/url_request/url_request_context_builder.h"
+ #include "net/url_request/url_request_context_getter.h"
+@@ -199,16 +201,33 @@ Cronet_RESULT Cronet_EngineImpl::StartWithParams(
+   // private and mark CronetLibraryLoader.postToInitThread() as
+   // @VisibleForTesting (as the only external use will be in a test).
+ 
++  // This supports a 'socks5' param to envoy:// URLs
++  GURL envoy_socks;
++  GURL envoy_temp = GURL(params->envoy_url);
++  for (net::QueryIterator it(envoy_temp); !it.IsAtEnd(); it.Advance()) {
++    auto key = it.GetKey();
++    auto value = it.GetUnescapedValue();
++    if (key.compare("socks5") == 0) {
++      envoy_socks =
++          GURL(base::UnescapeURLComponent(value, base::UnescapeRule::NORMAL));
++    }
++  }
+   // Initialize context on the init thread.
+-  if (params->envoy_url.rfind("socks5://", 0) != 0) {
+-  cronet::PostTaskToInitThread(
++  if (!envoy_socks.is_empty()) {
++    cronet::PostTaskToInitThread(
++      FROM_HERE,
++      base::BindOnce(&CronetContext::InitRequestContextOnInitThreadWithUri,
++                     base::Unretained(context_.get()), envoy_socks.spec()));
++  } else if (params->envoy_url.rfind("socks5://", 0) == 0) {
```

Comment:
was this intentionally changed from != to ==?

Reply:
Are you saying you think the logic is wrong?

Reply:
ah, i see now that the if/else blocks have been swapped too.